packages/xsnap/src/avaAssertXS.js

```diff
@@ -0,0 +1,351 @@
+  let beforeHooks = [];
+  /** @type {(() => Promise<void>)[]} */
+  let suitesToRun = [];
+  const context = {};
+
+  /**
+   * @returns { Summary }
+   * @typedef {import('./avaXS').Summary} Summary
+   */
+  function summary() {
+    return {
+      pass: passCount,
+      fail: testNum - passCount,
+      total: testNum,
+    };
+  }
+
+  const it = freeze({
+    send,
+    get context() {
+      return context;
+    },
+    /** @type {(l: string, f: () => Promise<void>) => void } */
+    before(_label, hook) {
+      beforeHooks.push(hook);
+    },
+    /** @type { (ok: boolean) => number } */
+    finish(ok) {
+      testNum += 1;
+      if (ok) {
+        passCount += 1;
+      }
+      return testNum;
+    },
+    /** @type { (f: () => Promise<void>) => Promise<void> } */
+    async defer(thunk) {
+      suitesToRun.push(thunk);
+    },
+    summary,
+    async result() {
+      for await (const hook of beforeHooks) {
+        await hook({ context });
+      }
+      beforeHooks = [];
+      for await (const suite of suitesToRun) {
+        await suite();
+      }
+      suitesToRun = [];
+
+      return summary();
+    },
+  });
+
+  if (!theHarness) {
+    theHarness = it;
+  }
+  return it;
+}
+
+/**
+ * @param {*} exc
+ * @param {Expectation} expectation
+ *
+ * @typedef {{ instanceOf: Function } | { message: string | RegExp }=} Expectation
+ */
+function checkExpectation(exc, expectation) {
+  if (!expectation) return true;
+  if ('instanceOf' in expectation) return exc instanceof expectation.instanceOf;
+  if ('message' in expectation) {
+    const { message } = expectation;
+    return typeof message === 'string'
+      ? exc.message === message
+      : exc.message.match(message);
+  }
+  throw Error(`not implemented: ${JSON.stringify(expectation)}`);
+}
+
+/**
+ * Emulate ava assertion API
+ *
+ * ref https://github.com/avajs/ava/blob/main/docs/03-assertions.md
+ *
+ * @param {Harness} htest
+ * @param {TapFormat} out
+ *
+ * @typedef {ReturnType<typeof makeTester>} Tester
+ */
+function makeTester(htest, out) {
+  /** @type {number?} */
+  let pending;
+
+  /** @type {(r: boolean, info?: string) => void} */
+  function assert(result, info) {
+    if (typeof pending === 'number') {
+      pending -= 1;
+    }
+    const testNum = htest.finish(result);
+    if (result) {
+      out.ok(testNum, info);
+    } else {
+      out.notOk(testNum, info);
+    }
+  }
+
+  function truthy(/** @type {unknown} */ value, msg = 'should be truthy') {
+    assert(!!value, msg);
+  }
+
+  /** @type {(a: unknown, e: unknown) => void } */
+  function deepEqTest(actual, expected) {
+    try {
+      assert(deepEqual(actual, expected), 'should be deep equal');
+    } catch (reason) {
+      const summary = JSON.stringify({ actual, expected });
+      assert(false, `should be deep equal: ${summary} : ${reason.message}`);
+    }
+  }
+
+  const t = freeze({
+    plan(/** @type {number} */ count) {
```

Comment:
I think I learned from @michaelfig that the type case syntax requires parens around the casted expression.
```suggestion
    plan(/** @type {number} */ (count)) {
```
and many other places

Reply:
this is not a cast. it's a parameter type declaration.